token.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2024 The Godror Authors
+//
+//
+// SPDX-License-Identifier: UPL-1.0 OR Apache-2.0
+
+package godror
+
+/*
+#include <stdlib.h>
+#include <stdio.h>
+#include "dpiImpl.h"
+
+int TokenCallbackHandlerDebug(void* context, dpiAccessToken *token);
+
+*/
+import "C"
+
+import (
+	"context"
+	"github.com/godror/godror/dsn"
+	"log"
+	"sync"
+	"unsafe"
+)
+
+// AccessToken Callback information.
+type accessTokenCB struct {
+	ctx         context.Context
+	callback    func(context.Context, *dsn.AccessToken) error
+	id          uint64
+	ctoken      *C.char
+	cprivateKey *C.char
+}
+
+// Cannot pass *accessTokenCB to C, so pass an uint64 that points to
+// this map entry
+var (
+	accessTokenCBsMu sync.Mutex
+	accessTokenCBs   = make(map[uint64]*accessTokenCB)
+	accessTokenCBsID uint64
+)
+
+// tokenCallbackHandler is the callback for C code on token expiry.
+//
+//export TokenCallbackHandler
+func TokenCallbackHandler(ctx unsafe.Pointer, accessTokenC *C.dpiAccessToken) {
+	log.Printf("CB %p %+v", ctx, accessTokenC)
+	accessTokenCBsMu.Lock()
+	acessTokenCB := accessTokenCBs[*((*uint64)(ctx))]
```

Comment:
if it is :=, then it overrides (shadows) the global accessTokenDBs, so no locking needed.

But I think you want to remove the :.

Reply:
I could not full get the comment..  I want to modify the value in the map with the token and private key C strings returned in this callback. These C strings are freed on next invocation of this callback.

I also modified to use RW lock instead of simple mutex. It looks ok?